internal/provider/group_resource.go

```diff
@@ -0,0 +1,366 @@
+package provider
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+        "strings"
+	"net/http"
+        "runtime"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+        "github.com/hashicorp/terraform-plugin-framework-jsontypes/jsontypes"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+    _ resource.Resource              = &groupResource{}
+)
+
+// NewGroupResource is a helper function to simplify the provider implementation.
+func NewGroupResource() resource.Resource {
+    return &groupResource{}
+}
+
+type GroupResourceModelInterface interface {
+	ParseHttpResponse(body []byte) error
+	CreateRequestBody() (*bytes.Reader, diag.Diagnostics)
+        GetURL() string
+}
+
+// groupResource is the resource implementation.
+type groupResource struct {
+    client ProviderHTTPClient
+}
+
+// Metadata returns the resource type name.
+func (r *groupResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+    resp.TypeName = req.ProviderTypeName + "_group"
+}
+
+// Schema defines the schema for the group resource.
+func (r *groupResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": schema.Int64Attribute{
+				Required: true,
+			},
+                        "inventory_id": schema.Int64Attribute{
+				Optional: true,
+			},
+			"name": schema.StringAttribute{
+				Required: true,
+			},
+                        "description": schema.StringAttribute{
+                                Optional: true,
+                                Computed:    true,
+                        },
+                        "group_url": schema.StringAttribute{
+				Computed: true,
+			},
+                        "variables": schema.StringAttribute{
+				Optional:   true,
+                                CustomType: jsontypes.NormalizedType{},
+			},
+		},
+	}
+}
+
+// groupResourceModel maps the resource schema data.
+type groupResourceModel struct {
+	Id    types.Int64  `tfsdk:"id"`
+        InventoryId   types.Int64  `tfsdk:"inventory_id"`
+	Name          types.String `tfsdk:"name"`
+	Description          types.String `tfsdk:"description"`
+        URL           types.String         `tfsdk:"group_url"`
+        Variables jsontypes.Normalized `tfsdk:"variables"`
+}
+
+func IsValueProvided(value attr.Value) bool {
+	return !value.IsNull() && !value.IsUnknown()
+}
+
+func (d *groupResourceModel) GetURL() string {
+        if !d.URL.IsNull() && !d.URL.IsUnknown() {
+		return d.URL.ValueString()
+	}
+	return ""
+}
+
+func GetFunctionName(caller int) string {
```

Comment:
Although this function returns the function name, the function name itself is meaningful and helps users identify the action where the error occurred. It was beneficial when distinguishing where the exception was thrown, either during the group reading or updating process. For instance, it would display something similar to "ReadGroup: Json Marshal Error." Do you believe this feature is still unnecessary?

Reply:
Yes, we should remove it. The internal function name shouldn't be part of the user facing error message. The context in which TF is being run will already tell the user if the error happened while creating, deleting, etc.